lib/CodeGen/Address.h

```diff
@@ -25,11 +25,41 @@ namespace CodeGen {
 class Address {
   llvm::Value *Pointer;
   CharUnits Alignment;
+
+private:
+  bool _containMMSafePtr;
+  llvm::Type *originalPointerTy;
+  llvm::Type *rawPointerTy;  // The inner pointer type of a _MMSafe_ptr.
+
 public:
   Address(llvm::Value *pointer, CharUnits alignment)
       : Pointer(pointer), Alignment(alignment) {
     assert((!alignment.isZero() || pointer == nullptr) &&
            "creating valid address with invalid alignment");
+    if (pointer) {
+      // Backup the original _MMSafe_ptr type.
+      originalPointerTy = pointer->getType();
+
+      // Checked C: for _MMSafe_ptr, reset the poitner type.
+      if (pointer->getType()->isMMSafePointerTy()) {
+        _containMMSafePtr = true;
+        rawPointerTy = pointer->getType()->getInnerPtrFromMMSafePtr();
+        pointer->mutateType(rawPointerTy);
+      }
+    }
+  }
+
+  // Return true if this Address contains a _MMSafe_ptr.
+  bool containMMSafePtr() const { return _containMMSafePtr; }
+
+  //  Set the pointer type to be the inner pointer type of a _MMSafe_ptr.
+  void mutatePointerType() {
+    if (containMMSafePtr()) Pointer->mutateType(rawPointerTy);
```

Comment:
Can Pointer be null? If not, consider adding an assertion.

Reply:
@sunnychatterjee Pointer here cannot be null  because when we need call this `mutatePointerType()` function, we've already known that we're using a non-void pointer in the program and that means the `Pointer` field has already been initialized to be non-null.